libmeepgeom/meepgeom.cpp

```diff
@@ -289,38 +250,52 @@ static geom_box gv2box(const meep::volume &v)
   return box;
 }
 
-// TODO rename this to something more descriptive like
-//      is_user_defined_material?
 static bool is_variable(material_type mt)
 {
-  return (mt->which_subclass == material_data::MATERIAL_FUNCTION);
+  return (mt->which_subclass == material_data::MATERIAL_USER);
 }
 static bool is_variable(void* md) { return is_variable((material_type) md); }
 
 static bool is_self(material_type md)
 {
-  return (md->which_subclass == material_data::MATERIAL_TYPE_SELF);
+  return (md->which_subclass == material_data::MATERIAL_FILE);
 }
 static bool is_self(void* md) { return is_self((material_type) md); }
 
 static bool is_medium(material_type md, medium_struct **m)
 {
   if (md->which_subclass == material_data::MEDIUM)
-   { *m = md->medium;
+   { *m = &(md->medium);
      return true;
    };
   return false;
 }
-static bool is_medium(void* md, medium_struct **m) { return is_medium((material_type) md, m); }
+
+static bool is_medium(void* md, medium_struct **m) 
+ { return is_medium((material_type) md, m); }
+
+static bool is_medium_or_user(material_type md, medium_struct **m)
+{ 
+  if (    md->which_subclass == material_data::MEDIUM
+       || md->which_subclass == material_data::MATERIAL_USER
+     )
+   { *m = &(md->medium);
```

Comment:
I don't understand how this works.  If it is `MATERIAL_USER`, don't you need to call the user function to get the material?

Reply:
The idea was that this function, like `material_epsmu`, should only ever be called *after* a call to `get_material_pt` and could thus rely on `md->medium` having been initialized.

But, to eliminate any assumptions, I modified `make_user_material` to call the user's function (at the origin) to fill in `md->medium` before returning the new `material_data` it creates. Thus, as long as user-defined materials are always created by calling `make_user_material,` the `md->medium` structure will always have been filled in by the user's function and will thus accurately reflect whether or not the material has nonlinearities (and, eventually, perhaps susceptibilities).